tests/expose_test.go

```diff
@@ -331,89 +557,172 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 			By("Getting the status of the VMI")
 			Eventually(func() bool {
-				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vm, err = virtClient.VirtualMachine(namespace).Get(name, &k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return vm.Status.Ready
 			}, 120*time.Second, 1*time.Second).Should(BeTrue())
 
 			By("Getting the running VMI")
 			var vmi *v1.VirtualMachineInstance
 			Eventually(func() bool {
-				vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vmi, err = virtClient.VirtualMachineInstance(namespace).Get(name, &k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return vmi.Status.Phase == v1.Running
 			}, 120*time.Second, 1*time.Second).Should(BeTrue())
 
-			tests.GenerateHelloWorldServer(vmi, testPort, "tcp")
-		})
+			return vmi
+		}
+
+		createStoppedVM := func(virtClient kubecli.KubevirtClient, namespace string) error {
+			By("Creating an VM object")
+			template := newLabeledVMI("vm", virtClient, false)
+			vm = tests.NewRandomVirtualMachine(template, false)
+
+			By("Creating the VM")
+			vm, err = virtClient.VirtualMachine(namespace).Create(vm)
+			return err
+		}
+
+		startVMWithServer := func(protocol string, port int) *v1.VirtualMachineInstance {
+			By("Calling the start command on the stopped VM")
+			vmi := startVMIFromVMTemplate(vm.GetName(), vm.GetNamespace())
+			if vmi == nil {
+				return nil
+			}
+			tests.GenerateHelloWorldServer(vmi, port, protocol)
+			return vmi
+		}
+
+		cleanupVM := func(virtClient kubecli.KubevirtClient, namespace string, vmName string) error {
+			By("Deleting the VM ...")
+			err := virtClient.VirtualMachine(vm.GetNamespace()).Delete(vm.GetName(), &k8smetav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+
+			err = wait.PollImmediate(time.Second, 2*time.Minute, func() (done bool, err error) {
+				_, err = virtClient.VirtualMachineInstance(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
+				return err != nil && errors.IsNotFound(err), nil
+			})
+			if err != nil {
+				return fmt.Errorf("error deleting VMI: %v", err)
+			}
+
+			err = wait.PollImmediate(time.Second, 2*time.Minute, func() (done bool, err error) {
+				_, err = virtClient.VirtualMachine(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
+				return err != nil && errors.IsNotFound(err), nil
+			})
+			if err != nil {
+				return fmt.Errorf("error deleting VM template: %v", err)
+			}
+			return err
+		}
 
 		Context("Expose a VM as a ClusterIP service.", func() {
-			It("[test_id:1538][label:masquerade_binding_connectivity]Connect to ClusterIP service that was set when VM was offline.", func() {
+			tests.BeforeAll(func() {
+				tests.BeforeTestCleanup()
+			})
+
+			BeforeEach(func() {
+				Expect(createStoppedVM(virtClient, tests.NamespaceTestDefault)).To(Succeed(), "should create a stopped VM.")
```

Comment:
i would suggest to add vm = nil as the first line
so it wont remember the vm which isnt nil from the previous cycle, in case the only line there fails.

Reply:
Why should I care about the pointer in the before each ?

Reply:
because you depend on it in the AfterEach and if the BeforeEach asserted, you would still have the previous value, and fail again on the assert of Delete in the AfterEach, since the ptr isnt nil but the vm is already deleted